### ISSUE TITLE
Connected TCP socket in image?

### ISSUE THREAD

**Issue by koliaskostas** [checkpoint-restore/criu#286]
After experimenting with the newer versions of CRIU I run into a problem when attempting to migrate.

First of all:

Purpose
--------
Migrate an apache service which is actively serving around 1,500 requests per second to a new host.

Setting
--------
Server:  Ubuntu 16.10 with the 4.10.0 kernel, CRIU 2.11,  apache2.4.18
Client: Ubuntu 16.10 with 4.8.0 kernel, running ab to create traffic

Commands
-----------
on the client: ab -n 500000 -c 100 http://192.168.56.102

on the server: criu dump --tree $APACHE_PID --images-dir $IMAGES_DIR --work-dir $WORK_DIR --log-file $LOG_DIR"log.dat" --skip-in-flight --tcp-established --leave-running -vvv

then I clone the server vm, and while apache is stopped I try to restore with:
restore --tree  $APACHE_PID --images-dir $IMAGES_DIR --work-dir  $WORK_DIR --log-file  $LOG_DIR"log.dat"-d -vvv

this results in errors:
(00.123498)   1374: Error (criu/sk-inet.c:654): Connected TCP socket in image
(00.124281)   1375: Error (criu/sk-inet.c:654): Connected TCP socket in image
(00.125242)   1371: Error (criu/cr-restore.c:1009): 1374 exited, status=1
(00.131067) Error (criu/cr-restore.c:1012): 1371 killed by signal 9: Killed
(00.131126) Error (criu/cr-restore.c:1897): Restoring FAILED.

Log
----
file
[log.data.zip](https://github.com/xemul/criu/files/797435/log.data.zip)

**Comment by xemul** [comment]
You should
  a) use --tcp-established option with restore command too
  b) make sure the network is locked the way it was by the dump command (netfilter rules or network-lock action script)

It's described in more details (I hope) here: https://criu.org/TCP_connection

**Comment by koliaskostas** [comment]
a) This flag was initially used but I got a previously discussed error message "Pid xxx do not match expected xxx" which back then was resolved by using the criu-ns executable from the scripts folder. This executable does not exist anymore. Attached is the log file when attempting to restore a (killed) apache process on the same machine. 
b) my tests were done either on the same vm or a clone of it. In all cases everything should be the same but the document is quite interesting. Would it be too much to ask to ament it with a minimal working example using network-lock action script?


[log.data.zip](https://github.com/xemul/criu/files/814156/log.data.zip)

**Comment by xemul** [comment]
@koliaskostas , the criu-ns script is still there, but it's available in criu-dev branch only. It still has issues we haven't resolved yet (mostly the lack of testing, but also issue #232) :( that's why we don't merge it into master and thus don't release. And yes, the "pid doesn't match expected" is all about pid conflicts that arise outside of pid namespaces.

As far as item b) is concerned, the generic example is hardly possible, as action scripts are designed to let others plug their own specific code, but fortunately, we have an example of how this works for LXC : https://github.com/xemul/criu/tree/master/test/others/app-emu/lxc . I can't promise this _does_ work with modern LXC, but the example you're looking for is: [this line](https://github.com/xemul/criu/blob/master/test/others/app-emu/lxc/run.sh#L41) that calls the network locking [script](https://github.com/xemul/criu/blob/master/test/others/app-emu/lxc/network-script.sh)  which, in turn, [locks](https://github.com/xemul/criu/blob/master/test/others/app-emu/lxc/network-script.sh#L40) or [unlocks](https://github.com/xemul/criu/blob/master/test/others/app-emu/lxc/network-script.sh#L34) the container's network.

**Comment by xemul** [comment]
@koliaskostas , is there anything else I can help you with this?

**Comment by xemul** [comment]
@koliaskostas , do you still have any concerns about this issue?

**Comment by koliaskostas** [comment]
@xemul So, I kept having this type of failures until I decided to manually add a iptables rule to cut out all http traffic (I know once again a sledgehammer) before I take a checkpoint. I will remind you my problems occurred  when under a flood of newly-created, short-lived connections. It worked for me with this trick but I haven't got to the bottom of the problem. Just mentioning this for people who are seeking something quick and dirty.
